editor/plugins/visual_shape_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,292 @@
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "visual_shape_2d_editor_plugin.h"
+
+#include "canvas_item_editor_plugin.h"
+#include "core/math/geometry_2d.h"
+#include "editor/editor_node.h"
+#include "editor/editor_undo_redo_manager.h"
+#include "editor/gui/editor_toaster.h"
+#include "editor/scene_tree_dock.h"
+#include "scene/2d/light_occluder_2d.h"
+#include "scene/2d/mesh_instance_2d.h"
+#include "scene/2d/physics/collision_shape_2d.h"
+#include "scene/2d/polygon_2d.h"
+#include "scene/gui/menu_button.h"
+#include "scene/resources/2d/capsule_shape_2d.h"
+#include "scene/resources/2d/circle_shape_2d.h"
+#include "scene/resources/2d/convex_polygon_shape_2d.h"
+#include "scene/resources/2d/rectangle_shape_2d.h"
+
+void VisualShape2DEditor::_node_removed(Node *p_node) {
+	if (p_node == visual_shape_2d) {
+		visual_shape_2d = nullptr;
+		options->hide();
+	}
+}
+
+void VisualShape2DEditor::edit(VisualShape2D *p_visual_shape_2d) {
+	visual_shape_2d = p_visual_shape_2d;
+}
+
+void VisualShape2DEditor::_menu_option(int p_option) {
+	if (!visual_shape_2d) {
+		return;
+	}
+
+	selected_menu_item = (Menu)p_option;
+
+	if (visual_shape_2d->get_owner() != get_tree()->get_edited_scene_root()) {
+		EditorToaster::get_singleton()->popup_str(TTR("Can't convert a VisualShape from a foreign scene."), EditorToaster::SEVERITY_ERROR);
```

Comment:
You forgot `return`.
Also creating shapes/occluders is not a conversion, so could be allowed.

Reply:
How can I check if the node is a scene root of a foreign scene?

Reply:
I meant root of the current scene. If `visual_shape_2d` is the root, it will have null owner, so this check fails.